### PR TITLE
Move create_collection over to new architecture

### PR DESCRIPTION
**Pull request by rylev**
This is an example of how to move a certain operation to the new pipeline based architecture. 

We want to do this for all operations.

### CODE REVIEW

File: sdk/cosmos/examples/document_00.rs
Comment (heaths) [inline sdk/cosmos/examples/document_00.rs:83]: Is this idiomatic to call `new()`, or should we implement the `Default` trait? I see that used in quite a few samples, though they are pretty basic.

File: sdk/core/src/prelude.rs
Comment (rylev) [reply sdk/core/src/prelude.rs:4]: I actually hope this _won't_ change often or at least not after some initial churn.

File: sdk/cosmos/examples/create_delete_database.rs
Comment (rylev) [reply sdk/cosmos/examples/create_delete_database.rs:71]: This is the [collection id](https://docs.microsoft.com/en-us/rest/api/cosmos-db/create-a-collection#body). I believe it's in the correct place here.

File: sdk/cosmos/examples/create_delete_database.rs
Comment (rylev) [reply sdk/cosmos/examples/create_delete_database.rs:72]: My understanding is that this is how partition keys are specified (with slashes). This is not an implementation detail.

cc @MindFlavor

File: sdk/cosmos/src/clients/database_client.rs
Comment (rylev) [reply sdk/cosmos/src/clients/database_client.rs:67]: Maybe? But I'm not sure if it would be worth the pain of tracking references across await points. We can explore this in another PR.

File: sdk/cosmos/examples/document_00.rs
Comment (rylev) [reply sdk/cosmos/examples/document_00.rs:83]: Implementing default is probably a good idea for this struct. Other options structs won't have defaults since they might have required options. Creating a collection, for example, requires a partition key (at least for now).

File: sdk/cosmos/src/clients/database_client.rs
Comment (rylev) [reply sdk/cosmos/src/clients/database_client.rs:67]: Maybe? But I'm not sure if it would be worth the pain of tracking references across await points. We can explore this in another PR.

File: sdk/cosmos/src/operations/create_collection.rs
Comment (rylev) [reply sdk/cosmos/src/operations/create_collection.rs:9]: Understand regarding wanting the name to unambiguous when unqualified. 👍

However, I thought that the options were operation specific not client specific. In other words, creating a collection will take a different options type than deleting the collection. Is that not the case?

Comment (heaths) [reply sdk/cosmos/src/operations/create_collection.rs:9]: Since the API is still in flux (old style vs. moving to Azure SDK guidelines) I wasn't sure, but if it's a per-method options, we *usually* follow <MethodName>Options so the options type is unambiguously tied to the actual method name. If it's shared by multiple methods, I would still qualify it but it could be something matching closer to the client.

File: sdk/cosmos/examples/create_delete_database.rs
Comment (rylev) [reply sdk/cosmos/examples/create_delete_database.rs:69]: Discussed this offline with @JeffreyRichter. The Context argument is actually the first argument in C++ and Go. I _personally_ am a fan of it being last, but we decided to open an issue to discuss this instead of blocking this PR.

**Comment by rylev** [comment]
@heaths I think this is good to go.

**Review by heaths — APPROVED** [review]

File: sdk/cosmos/src/clients/cosmos_client.rs
Comment (MindFlavor) [inline sdk/cosmos/src/clients/cosmos_client.rs:166]: Why this new name? I prefer as it was before, just `Response`. 
With the module we leave choice to the SDK user: they can either write `Response` or `create_database::Response`. With `CreateDatabaseResponse` we force the verbosity on them.

File: sdk/cosmos/src/clients/database_client.rs
Comment (MindFlavor) [inline sdk/cosmos/src/clients/database_client.rs:58]: Same as above, I think `create_collection::Options` would be less verbose and flexible.

File: sdk/cosmos/src/clients/database_client.rs
Comment (MindFlavor) [reply sdk/cosmos/src/clients/database_client.rs:67]: I agree, I do not think avoiding a clone is worth the lifetime trouble. If `Context` becomes expensive to clone we might consider reference counting?

**Review by MindFlavor — CHANGES_REQUESTED** [review]
Great work. 👍 

I am skeptical about the naming change that made the code much more verbose. Are we sure we want to follow this standard? We will end up with `GetContainerACLOptions`, `GetContainerACLResponse` etc...  that are painful to read. 

What about the previous standard (use modules)? I think is more idiomatic in Rust.

File: sdk/cosmos/src/clients/cosmos_client.rs
Comment (heaths) [reply sdk/cosmos/src/clients/cosmos_client.rs:166]: See my [comment](https://github.com/Azure/azure-sdk-for-rust/pull/279#discussion_r643558552). Namespacing doesn't really matter. If you want to make the case it's idiomatic, I defer to @JeffreyRichter but these naming conventions (namespaced or not) have been a point of contention for a while.

Comment (heaths) [reply sdk/cosmos/src/clients/cosmos_client.rs:166]: Unless something truly is idiomatic, we should strive to be consistent. We are across all the SDKs by design. The idea is to make familiarity with one language SDK portable to another and open up multi-project, multi-lang solutions for all devs targeting Azure.

Comment (JeffreyRichter) [reply sdk/cosmos/src/clients/cosmos_client.rs:166]: Other SDKs return a Response<T> where T is the XxxResult where Xxx is the name of the operation (like CreateDatabaseResult). So, we might want to return a Response<CreateDatabaseResult>. I know this is long but it is what other languages do. 

A Response<T> has the RawResponse field in it which has the HTTP status code and the body string (as downloaded from the service) -- customers typically ignore this entirely but it is very useful for debugging scenarios. Also, the Response has the T which is for the operation's model for the deserialized result (XxxResult).

The XxxResult has fields in it for every deserialized response header and another field for the Body; the type of the Body field is the model that is specific to the operation. This contains the deserialized resource returned from the service. The body is frequently sent back to the service. For example, the customer GETs a resource, modifies the Body model and then PUTs or PATCHes the resource back up to the service. Response headers are never sent back to the service.

Comment (MindFlavor) [reply sdk/cosmos/src/clients/cosmos_client.rs:166]: Ah, ok, that weird Java feeling I had was on purpose!😄

Collapsing all the options and responses in a module (like it is now in [operations/mod.rs](https://github.com/Azure/azure-sdk-for-rust/pull/279/files#diff-3a33a87e7d35109d7ee3489b6cb76e1cd3d61c2112643d6e1728b6a9f4fe1c07)) will create a very poor autocompletion experience (you'll either import the types by hand or end up with a ton of structs, at least two for each operation). 

That said, since it has already been discussed, I'll resolve this conversation.

**Review by MindFlavor — APPROVED** [review]

**Comment by rylev** [comment]
@MindFlavor I was also concerned about verbosity, but I'd like to give this way a try. It doesn't seem too bad right now.

**Pull request merged**
